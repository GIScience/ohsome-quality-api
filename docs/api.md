# API

Please have a look at the request examples of the [interactive Swagger UI](https://oqt.ohsome.org/api/docs).
Using the Swagger UI one can create valid `cURL` requests with using an interactive interface.


## Indicator Endpoint

Three types of requests can be made:

- To request an Indicator for an AOI defined by OQT provide the following parameters: `name`, `layerKey`, `dataset` and `featureId`
- To request an Indicator for a custom AOI provide the following parameters: parameters: `name`, `layerKey` and `bpolys`
- To request an Indicator for a custom AOI and custom data provide the following parameters: `name`, `bpolys` and `layer`

Depending on the input, the output will be a GeoJSON Feature or FeatureCollection with the results of the Indicator as properties. The Feature properties of the input GeoJSON will be preserved if they do not collide with the properties set by OQT.

### Parameters

| Parameter     | Description                                                                                                                                       |
| ---           | ---                                                                                                                                               |
| `name`        | Name of the Indicator                                                                                                                             |
| `layerKey`    | Name of the layer definition                                                                                                                      |
| `dataset`     | Name of the dataset containing bounding polygons                                                                                                  |
| `featureId`   | Identifier of the feature in the dataset                                                                                                          |
| `bpolys`      | Bounding polygon(s) as GeoJSON Feature, FeatureCollection, Polygon or MultiPolygon object. The Geometry has to be of type Polygon or MultiPolygon |
| `layer`       | Name, description and data of a custom layer. Data has to be provided in the same format as the response of the ohsome API would look like        |
| `includeSvg`  | Include a SVG string of a figure displaying the Indicator results                                                                                 |
| `includeHtml` | Include a HTML string with the Indicator results                                                                                                  |
| `flatten`     | Flattening of the GeoJSON properties                                                                                                              |

Parameters are in part mutual exclusive (e.g. either `dataset` and `featureId` or `bpolys`). If the response contains multiple errors probably not all required parameters and/or mutual exclusive parameters have been provided.

### A Note on the Difference of `GET` and `POST` Requests

A request can be made to the `/indicator` endpoint using either the `GET` or `POST` method. For `GET` requests to this endpoint, only the parameters `dataset` and `feature_id` are supported. This means that it is not possible to create an indicator for a custom AOI using the `bpolys` parameter by making a `GET` request to the API.

Using the `POST` method all types of requests (see above) and parameters are supported.

### Request an Indicator for an OQT AOI using Python and `requests` Library

```python
import requests

url = "https://oqt.ohsome.org/api/indicator"
parameters = {
    "name": "GhsPopComparisonBuildings",
    "layerKey": "building_count",
    "dataset": "regions",
    "featureId": 3,
    "fidField": "ogc_fid",  # Optional
    "includeSvg": False,  # Optional
    "includeHtml": False,  # Optional
    "flatten": False,  # Optional
}
# Response using the GET method
response = requests.get(url, params=parameters)
# Response using the POST method
response = requests.post(url, json=parameters)
```

### Request an Indicator for a custom AOI using Python and `requests` Library

```python
import requests

url = "https://oqt.ohsome.org/api/indicator"
bpolys = {
    "type": "Polygon",
    "coordinates": [
        [
            [8.674092292785645, 49.40427147224242],
            [8.695850372314453, 49.40427147224242],
            [8.695850372314453, 49.415552187316095],
            [8.674092292785645, 49.415552187316095],
            [8.674092292785645, 49.40427147224242],
        ]
    ],
}
parameters = {
    "name": "GhsPopComparisonBuildings",
    "layerKey": "building_count",
    "bpolys": bpolys,
    "includeSvg": False,  # Optional
    "includeHtml": False,  # Optional
    "flatten": False,  # Optional
}
response = requests.post(url, json=parameters)
```

### Request an Indicator for a custom AOI and Layer using Python and `requests` Library

The data attached to a custom layer has to be in the same structure as the ohsome API response: [https://docs.ohsome.org/ohsome-api/stable](https://docs.ohsome.org/ohsome-api/stable)

Data used in this example has been taken from the ohsome API response of following request URL: [https://api.ohsome.org/v1/elements/count?bboxes=8.67%2C49.39%2C8.71%2C49.42&filter=building%3D\*%20and%20geometry%3Apolygon&format=json&time=2014-01-01%2F2017-01-01%2FP1M](https://api.ohsome.org/v1/elements/count?bboxes=8.67%2C49.39%2C8.71%2C49.42&filter=building%3D*%20and%20geometry%3Apolygon&format=json&time=2014-01-01%2F2017-01-01%2FP1M)

```python
import requests

url = "https://oqt.ohsome.org/api/indicator"
bpolys = {
    "type": "Polygon",
    "coordinates": [
        [
            [8.674092292785645, 49.40427147224242],
            [8.695850372314453, 49.40427147224242],
            [8.695850372314453, 49.415552187316095],
            [8.674092292785645, 49.415552187316095],
            [8.674092292785645, 49.40427147224242],
        ]
    ],
}
layer = {
    "name": "My layer name"
    "description": "My layer description"
    "data": {
        "result": [
            {"timestamp": "2014-01-01T00:00:00Z", "value": 4708},
            {"timestamp": "2014-02-01T00:00:00Z", "value": 4842},
            {"timestamp": "2014-03-01T00:00:00Z", "value": 4840},
            {"timestamp": "2014-04-01T00:00:00Z", "value": 4941},
            {"timestamp": "2014-05-01T00:00:00Z", "value": 4987},
            {"timestamp": "2014-06-01T00:00:00Z", "value": 5007},
            {"timestamp": "2014-07-01T00:00:00Z", "value": 5020},
            {"timestamp": "2014-08-01T00:00:00Z", "value": 5168},
            {"timestamp": "2014-09-01T00:00:00Z", "value": 5355},
            {"timestamp": "2014-10-01T00:00:00Z", "value": 5394},
            {"timestamp": "2014-11-01T00:00:00Z", "value": 5449},
            {"timestamp": "2014-12-01T00:00:00Z", "value": 5470},
            {"timestamp": "2015-01-01T00:00:00Z", "value": 5475},
            {"timestamp": "2015-02-01T00:00:00Z", "value": 5477},
            {"timestamp": "2015-03-01T00:00:00Z", "value": 5481},
            {"timestamp": "2015-04-01T00:00:00Z", "value": 5495},
            {"timestamp": "2015-05-01T00:00:00Z", "value": 5516},
            {"timestamp": "2015-06-01T00:00:00Z", "value": 5517},
            {"timestamp": "2015-07-01T00:00:00Z", "value": 5519},
            {"timestamp": "2015-08-01T00:00:00Z", "value": 5525},
            {"timestamp": "2015-09-01T00:00:00Z", "value": 5560},
            {"timestamp": "2015-10-01T00:00:00Z", "value": 5564},
            {"timestamp": "2015-11-01T00:00:00Z", "value": 5568},
            {"timestamp": "2015-12-01T00:00:00Z", "value": 5627},
            {"timestamp": "2016-01-01T00:00:00Z", "value": 5643},
            {"timestamp": "2016-02-01T00:00:00Z", "value": 5680},
            {"timestamp": "2016-03-01T00:00:00Z", "value": 5681},
            {"timestamp": "2016-04-01T00:00:00Z", "value": 5828},
            {"timestamp": "2016-05-01T00:00:00Z", "value": 5974},
            {"timestamp": "2016-06-01T00:00:00Z", "value": 5990},
            {"timestamp": "2016-07-01T00:00:00Z", "value": 5991},
            {"timestamp": "2016-08-01T00:00:00Z", "value": 5997},
            {"timestamp": "2016-09-01T00:00:00Z", "value": 6002},
            {"timestamp": "2016-10-01T00:00:00Z", "value": 6010},
            {"timestamp": "2016-11-01T00:00:00Z", "value": 6010},
            {"timestamp": "2016-12-01T00:00:00Z", "value": 6016},
            {"timestamp": "2017-01-01T00:00:00Z", "value": 6015},
        ]
    }
}
parameters = {
    "name": "GhsPopComparisonBuildings",
    "bpolys": bpolys,
    "layer": layer,
    "includeSvg": False,  # Optional
    "includeHtml": False,  # Optional
    "flatten": True,  # Optional
}
response = requests.post(url, json=parameters)
```

### Response Examples

```json
{
  "apiVersion": "0.7.0",
  "attribution": {
    "text": "Â© OpenStreetMap contributors",
    "url": "https://ohsome.org/copyrights"
  },
  "type": "Feature",
  "geometry": {
    "type": "MultiPolygon",
    "coordinates": [ ... ]
  },
  "properties": {
    "metadata.name": "GHS-POP Comparison Buildings",
    "metadata.description": "Comparison between population density and feature density.\nThis can give an estimate if mapping has been completed.\n",
    "layer.name": "Building Count",
    "layer.description": "All buildings as defined by all objects tagged with 'building=*'.\n",
    "result.timestamp_oqt": "2021-10-05T09:33:30.671060+00:00",
    "result.timestamp_osm": "2021-09-26T20:00:00+00:00",
    "result.label": "green",
    "result.value": 1,
    "result.description": "...",
    "result.svg": "...",
    "data.pop_count": 533351.7174390131,
    "data.area": 70.29342423663759,
    "data.pop_count_per_sqkm": 7587.5051362347085,
    "data.feature_count": 69993,
    "data.feature_count_per_sqkm": 995.7261402485354
  }
}
```


## Report Endpoint

Two types of requests can be made:

- To request a Report for an AOI defined by OQT provide the following parameters: `name`, `dataset` and `featureId`
- To request a Report for a custom AOI provide the following parameters: parameters: `name` and `bpolys`

Depending on the input, the output will be a GeoJSON Feature or FeatureCollection with the results of the Report and each Indicator as properties. The Feature properties of the input GeoJSON will be preserved if they do not collide with the properties set by OQT.

### Parameters

| Parameter     | Description                                                                                                                                       |
| ---           | ---                                                                                                                                               |
| `name`        | Name of the Report                                                                                                                                |
| `dataset`     | Name of the dataset containing bounding polygons                                                                                                  |
| `featureId`   | Identifier of the feature in the dataset                                                                                                          |
| `bpolys`      | Bounding polygon(s) as GeoJSON Feature, FeatureCollection, Polygon or MultiPolygon object. The Geometry has to be of type Polygon or MultiPolygon |
| `includeSvg`  | Include a SVG string of a figure displaying the Indicator results                                                                                 |
| `includeHtml` | Include a HTML string with the Indicator results                                                                                                  |
| `flatten`     | Flattening of the GeoJSON properties                                                                                                              |

Some parameters are mutually exclusive (E.g. either `dataset` and `featureId` or `bpolys`). If the response contains multiple errors, it is likely that not all required parameters and/or mutually exclusive parameters have been provided.

### A Note on the Difference of `GET` and `POST` Requests

A request can be made to the `/report` endpoint using either the `GET` or `POST` method. For `GET` requests to this endpoint, only the `dataset` and `feature_id` parameters are supported. That means that it is not possible to create a Report for a custom AOI using the `bpolys` parameter by making a `GET` request to the API.

Using the `POST` method all types of requests (see above) and parameters are supported.

### Request an Report for an OQT AOI using Python and `requests` Library

```python
import requests

url = "https://oqt.ohsome.org/api/report"
parameters = {
    "name": "MinimalTestReport",
    "dataset": "regions",
    "featureId": 3,
    "includeSvg": False,  # Optional
    "includeHtml": False,  # Optional
    "flatten": False,  # Optional
}
# Response using the GET method
response = requests.get(url, params=parameters)
# Response using the POST method
response = requests.post(url, json=parameters)
```

### Request an Report for a custom AOI using Python and `requests` Library

```python
import requests

url = "https://oqt.ohsome.org/api/report"
bpolys = {
    "type": "Polygon",
    "coordinates": [
        [
            [8.674092292785645, 49.40427147224242],
            [8.695850372314453, 49.40427147224242],
            [8.695850372314453, 49.415552187316095],
            [8.674092292785645, 49.415552187316095],
            [8.674092292785645, 49.40427147224242],
        ]
    ],
}
parameters = {
    "name": "MinimalTestReport",
    "bpolys": bpolys,
    "includeSvg": False,  # Optional
    "includeHtml": False,  # Optional
    "includeData": False,  # Optional
    "flatten": False,  # Optional
}
response = requests.post(url, json=parameters)
```

### Response Examples

```json
{
  "apiVersion": "0.10.0",
  "attribution": {
    "url": "https://github.com/GIScience/ohsome-quality-analyst/blob/main/data/COPYRIGHTS.md",
    "text": "Â© OpenStreetMap contributors"
  },
  "type": "Feature",
  "geometry": {
    "type": "MultiPolygon",
    "coordinates": [ ... ]
  },
  "properties": {
    "report.metadata.name": "Minimal",
    "report.metadata.description": "This report shows the quality for two indicators:\nMapping Saturation and Currentness.\nIt's main function is to test the interactions between\ndatabase, api and website.\n",
    "report.result.label": "green",
    "report.result.value": 0.8902104187552007,
    "report.result.description": "All indicators show a good quality.\nThe data in this regions seems to be completely mapped.\n",
    "indicators.0.metadata.name": "Mapping Saturation",
    "indicators.0.metadata.description": "Mapping Saturation",
    "indicators.0.layer.name": "Building Count",
    "indicators.0.layer.description": "All buildings as defined by all objects tagged with 'building=*'.\n",
    "indicators.0.result.timestamp_oqt": "2022-07-10T15:18:05.208875+00:00",
    "indicators.0.result.timestamp_osm": "2022-03-01T00:00:00+00:00",
    "indicators.0.result.label": "green",
    "indicators.0.result.value": 0.9804208375104013,
    "indicators.0.result.description": "The saturation of the last 3 years is 98.04%.\nHigh saturation has been reached (97% < Saturation â¤ 100%).\n",
    "indicators.1.metadata.name": "Currentness",
    "indicators.1.metadata.description": "Currentness",
    "indicators.1.layer.name": "Building Count",
    "indicators.1.layer.description": "All buildings as defined by all objects tagged with 'building=*'.\n",
    "indicators.1.result.timestamp_oqt": "2022-07-10T15:33:36.449060+00:00",
    "indicators.1.result.timestamp_osm": "2022-07-03T20:00:00+00:00",
    "indicators.1.result.label": "green",
    "indicators.1.result.value": 0.8,
    "indicators.1.result.description": "Over 50% of the 40811.0 features (Building Count) were edited in the last 4 years.\nThis is a rather high value and indicates that the map features \nare very unlike to be outdated. This refers to good data quality in \nrespect to currentness.\n"
  }
}
```
