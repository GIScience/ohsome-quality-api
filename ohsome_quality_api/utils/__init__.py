from .pybabel_yaml_extractor import pybabel_yaml_extractor

__all__ = ("pybabel_yaml_extractor",)
