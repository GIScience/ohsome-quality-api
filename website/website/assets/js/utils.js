function get_html_parameter_list(querystring) {
    if (querystring === '') return {};
    const params = querystring.slice(1);
    const pairs = params.split("&");
    const result = {}
    for (const item of pairs) {
      const pair = item.split("=");
      const name = unescape(pair[0]).replace("+", " ");
      const value = unescape(pair[1]).replace("+", " ");
      result[name] = value;
    }
    return result
  }

function update_url(key, value){
    const params = get_html_parameter_list(location.search)
    params[key] = value
    let new_params = "?"
    for (const x of Object.keys(params)) {
        new_params += x + "=" + params[x] + "&"
    }
    new_params = new_params.slice(0,-1)
    window.history.replaceState(null, null, new_params);
}

// check if a report with this name is implemented
function report_isValid(report){
    const valid_options = []
    for (const option of document.getElementById("cardtype").options) {
        valid_options.push(option.value)
      }
    return valid_options.includes(report)
}

// check if test region id is in the test regions file. These functions need to be updated once we have other inputs.
function id_isValid(value, json){
    const valid_options = []
    for (const elem of json){
        valid_options.push(elem.id)
    }
    return valid_options.includes(value)
}
