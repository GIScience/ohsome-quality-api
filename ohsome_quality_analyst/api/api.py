import json
import logging
import os
from typing import Annotated

from fastapi import FastAPI, HTTPException, Path, Request, status
from fastapi.encoders import jsonable_encoder
from fastapi.exceptions import RequestValidationError
from fastapi.middleware.cors import CORSMiddleware
from fastapi.openapi.docs import (
    get_redoc_html,
    get_swagger_ui_html,
    get_swagger_ui_oauth2_redirect_html,
)
from fastapi.responses import JSONResponse
from starlette.exceptions import HTTPException as StarletteHTTPException
from starlette.staticfiles import StaticFiles

from ohsome_quality_analyst import (
    __author__,
    __description__,
    __email__,
    __homepage__,
    __title__,
    __version__,
    oqt,
)
from ohsome_quality_analyst.api.request_models import (
    IndicatorDataRequest,
    IndicatorEnum,
    IndicatorRequest,
    ReportEnum,
    ReportRequest,
    TopicEnum,
)
from ohsome_quality_analyst.api.response_models import (
    IndicatorMetadataResponse,
    MetadataResponse,
    ProjectMetadataResponse,
    QualityDimensionMetadataResponse,
    ReportMetadataResponse,
    TopicMetadataResponse,
)
from ohsome_quality_analyst.config import configure_logging
from ohsome_quality_analyst.definitions import (
    ATTRIBUTION_URL,
    get_indicator_definitions,
    get_metadata,
    get_report_definitions,
    get_topic_definition,
    get_topic_definitions,
)
from ohsome_quality_analyst.projects.definitions import (
    ProjectEnum,
    get_project,
    get_projects,
)
from ohsome_quality_analyst.quality_dimensions.definitions import (
    QualityDimensionEnum,
    get_quality_dimension,
    get_quality_dimensions,
)
from ohsome_quality_analyst.utils.exceptions import (
    HexCellsNotFoundError,
    OhsomeApiError,
    RasterDatasetNotFoundError,
    RasterDatasetUndefinedError,
    SizeRestrictionError,
    TopicDataSchemaError,
    ValidationError,
)
from ohsome_quality_analyst.utils.helper import (
    get_class_from_key,
    hyphen_to_camel,
    json_serialize,
)
from ohsome_quality_analyst.utils.validators import validate_indicator_topic_combination

MEDIA_TYPE_GEOJSON = "application/geo+json"
MEDIA_TYPE_JSON = "application/json"
STATIC_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "static")

TAGS_METADATA = [
    {
        "name": "indicator",
        "description": "Request an Indicator",
        "externalDocs": {
            "description": "External docs",
            "url": (
                "https://github.com/GIScience/ohsome-quality-analyst/blob/"
                + __version__
                + "/docs/api.md"
            ),
        },
    },
    {
        "name": "report",
        "description": "Request a Report",
        "externalDocs": {
            "description": "External docs",
            "url": (
                "https://github.com/GIScience/ohsome-quality-analyst/blob/"
                + __version__
                + "/docs/api.md"
            ),
        },
    },
    {
        "name": "metadata",
        "description": "Request Metadata",
        "externalDocs": {
            "description": "External docs",
            "url": (
                "https://github.com/GIScience/ohsome-quality-analyst/blob/"
                + __version__
                + "/docs/api.md"
            ),
        },
    },
]

# TODO: to be replaced by config
DEFAULT_PROJECT = ProjectEnum.core

configure_logging()
logging.info("Logging enabled")
logging.debug("Debugging output enabled")

app = FastAPI(
    title=__title__,
    description=__description__,
    version=__version__,
    contact={
        "name": __author__,
        "url": __homepage__,
        "email": __email__,
    },
    openapi_tags=TAGS_METADATA,
    docs_url=None,
    redoc_url=None,
)

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
app.mount("/static", StaticFiles(directory=STATIC_DIR), name="static")


@app.get("/docs", include_in_schema=False)
async def custom_swagger_ui_html(request: Request):
    root_path = request.scope.get("root_path")
    return get_swagger_ui_html(
        openapi_url=root_path + app.openapi_url,
        title=app.title + " - Swagger UI",
        oauth2_redirect_url=app.swagger_ui_oauth2_redirect_url,
        swagger_js_url=root_path + "/static/swagger-ui-bundle.js",
        swagger_css_url=root_path + "/static/swagger-ui.css",
        swagger_favicon_url=root_path + "/static/favicon-32x32.png",
    )


@app.get(app.swagger_ui_oauth2_redirect_url, include_in_schema=False)
async def swagger_ui_redirect():
    return get_swagger_ui_oauth2_redirect_html()


@app.get("/redoc", include_in_schema=False)
async def redoc_html(request: Request):
    root_path = request.scope.get("root_path")
    return get_redoc_html(
        openapi_url=root_path + app.openapi_url,
        title=app.title + " - ReDoc",
        redoc_js_url=root_path + "/static/redoc.standalone.js",
        redoc_favicon_url=root_path + "/static/favicon-32x32.png",
    )


class CustomJSONResponse(JSONResponse):
    def render(self, content):
        return json.dumps(content, default=json_serialize).encode()


@app.exception_handler(RequestValidationError)
async def validation_exception_handler(
    _: Request,
    exception: RequestValidationError,
):
    """Exception handler for validation exceptions."""
    return JSONResponse(
        status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
        content=jsonable_encoder(
            {
                "apiVersion": __version__,
                "type": exception.__class__.__name__,
                "detail": exception.errors(),
            },
        ),
    )


@app.exception_handler(HexCellsNotFoundError)
@app.exception_handler(TopicDataSchemaError)
@app.exception_handler(OhsomeApiError)
@app.exception_handler(RasterDatasetNotFoundError)
@app.exception_handler(RasterDatasetUndefinedError)
@app.exception_handler(SizeRestrictionError)
@app.exception_handler(ValidationError)
async def oqt_exception_handler(
    _: Request,
    exception: HexCellsNotFoundError
    | TopicDataSchemaError
    | OhsomeApiError
    | RasterDatasetNotFoundError
    | RasterDatasetUndefinedError
    | SizeRestrictionError
    | ValidationError,
):
    """Exception handler for OQT exceptions."""
    return JSONResponse(
        status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
        content={
            "apiVersion": __version__,
            "type": exception.name,
            "detail": [
                {
                    "msg": exception.message,
                },
            ],
        },
    )


@app.exception_handler(StarletteHTTPException)
async def http_exception_handler(_: Request, exception: StarletteHTTPException):
    return JSONResponse(
        status_code=exception.status_code,
        content={
            "apiVersion": __version__,
            "type": exception.__class__.__name__,
            "detail": [
                {
                    "msg": exception.detail,
                },
            ],
        },
    )


def empty_api_response() -> dict:
    return {
        "apiVersion": __version__,
        "attribution": {
            "url": ATTRIBUTION_URL,
        },
    }


@app.post("/indicators/mapping-saturation/data", include_in_schema=False)
async def post_indicator_ms(parameters: IndicatorDataRequest) -> CustomJSONResponse:
    """Legacy support for computing the Mapping Saturation indicator for given data."""
    geojson_object = await oqt.create_indicator_as_geojson(
        parameters,
        key="mapping-saturation",
    )
    response = empty_api_response()
    response["attribution"]["text"] = get_class_from_key(
        class_type="indicator",
        key="mapping-saturation",
    ).attribution()
    # TODO: if accept=JSON no GeoJSON should be created in the first place.
    #   factor out logic and decision to base/indicator.py and oqt.py
    #   base/indicator.py should have `as_dict` alongside `as_feature`
    response["results"] = [feature.properties for feature in geojson_object.features]
    return CustomJSONResponse(content=response, media_type=MEDIA_TYPE_JSON)


@app.post(
    "/indicators/{key}",
    tags=["indicator"],
    responses={
        200: {
            "content": {MEDIA_TYPE_GEOJSON: {}},
            "description": "Return JSON or GeoJSON.",
        }
    },
)
async def post_indicator(
    request: Request,
    key: Annotated[
        IndicatorEnum,
        Path(
            title="Indicator Key",
            example="mapping-saturation",
        ),
    ],
    parameters: IndicatorRequest,
) -> CustomJSONResponse:
    """Request an Indicator for an AOI defined by OQT or a custom AOI."""
    if isinstance(parameters, IndicatorRequest):
        validate_indicator_topic_combination(key.value, parameters.topic_key.value)
    geojson_object = await oqt.create_indicator_as_geojson(parameters, key=key.value)
    response = empty_api_response()
    response["attribution"]["text"] = get_class_from_key(
        class_type="indicator",
        key=key.value,
    ).attribution()
    # TODO: if accept=JSON no GeoJSON should be created in the first place.
    #   factor out logic and decision to base/indicator.py and oqt.py
    #   base/indicator.py should have `as_dict` alongside `as_feature`
    if request.headers["accept"] == MEDIA_TYPE_JSON:
        response["results"] = [
            feature.properties for feature in geojson_object.features
        ]
        return CustomJSONResponse(content=response, media_type=MEDIA_TYPE_JSON)
    elif request.headers["accept"] == MEDIA_TYPE_GEOJSON:
        response.update(geojson_object)
        return CustomJSONResponse(content=response, media_type=MEDIA_TYPE_GEOJSON)
    else:
        detail = "Content-Type needs to be either {0} or {1}".format(
            MEDIA_TYPE_JSON, MEDIA_TYPE_GEOJSON
        )
        raise HTTPException(
            status_code=status.HTTP_415_UNSUPPORTED_MEDIA_TYPE, detail=detail
        )


@app.post("/reports/{key}", tags=["report"])
async def post_report(
    key: Annotated[
        ReportEnum,
        Path(
            title="Report Key",
            example="building-report",
        ),
    ],
    parameters: ReportRequest,
) -> CustomJSONResponse:
    """Request a Report for an AOI defined by OQT or a custom AOI."""
    geojson_object = await oqt.create_report_as_geojson(parameters, key=key.value)
    response = empty_api_response()
    response["attribution"]["text"] = get_class_from_key(
        class_type="report",
        key=key.value,
    ).attribution()
    response.update(geojson_object)
    return CustomJSONResponse(content=response, media_type=MEDIA_TYPE_GEOJSON)


@app.get(
    "/metadata",
    tags=["metadata"],
    response_model_exclude={
        "result": {
            "topics": {k.value: {"key": True} for k in TopicEnum},
            "indicators": {
                k.value: {"label_description": True, "result_description": True}
                for k in IndicatorEnum
            },
            "reports": {k.value: {"label_description": True} for k in ReportEnum},
        }
    },
)
async def metadata(project: ProjectEnum = DEFAULT_PROJECT) -> MetadataResponse:
    """Get topics."""
    if project == ProjectEnum.all:
        project = None
    result = {
        "topics": get_topic_definitions(project=project),
        "quality_dimensions": get_quality_dimensions(),
        "projects": get_projects(),
        "indicators": get_indicator_definitions(project=project),
        "reports": get_report_definitions(project=project),
    }
    return MetadataResponse(result=result)


@app.get(
    "/metadata/topics",
    tags=["metadata"],
    response_model_exclude={
        "result": {k.value: {"key": True} for k in TopicEnum},
    },
)
async def metadata_topic(
    project: ProjectEnum = DEFAULT_PROJECT,
) -> TopicMetadataResponse:
    """Get topics."""
    if project == ProjectEnum.all:
        project = None
    return TopicMetadataResponse(result=get_topic_definitions(project=project))


@app.get(
    "/metadata/topics/{key}",
    tags=["metadata"],
    response_model_exclude={"result": {k.value: {"key": True} for k in TopicEnum}},
)
async def metadata_topic_by_key(key: TopicEnum) -> TopicMetadataResponse:
    """Get topic by key."""
    return TopicMetadataResponse(result={key.value: get_topic_definition(key.value)})


@app.get(
    "/metadata/quality-dimensions",
    tags=["metadata"],
)
async def metadata_quality_dimensions() -> QualityDimensionMetadataResponse:
    """Get quality dimensions."""
    return QualityDimensionMetadataResponse(result=get_quality_dimensions())


@app.get(
    "/metadata/quality-dimensions/{key}",
    tags=["metadata"],
)
async def metadata_quality_dimension_by_key(
    key: QualityDimensionEnum,
) -> QualityDimensionMetadataResponse:
    """Get quality dimension by key."""
    return QualityDimensionMetadataResponse(
        result={key.value: get_quality_dimension(key.value)}
    )


@app.get(
    "/metadata/projects",
    tags=["metadata"],
)
async def metadata_projects() -> ProjectMetadataResponse:
    """Get projects."""
    return ProjectMetadataResponse(result=get_projects())


@app.get(
    "/metadata/projects/{key}",
    tags=["metadata"],
)
async def metadata_project_by_key(
    key: ProjectEnum,
) -> ProjectMetadataResponse:
    """Get project by key."""
    return ProjectMetadataResponse(result={key.value: get_project(key.value)})


@app.get(
    "/metadata/indicators",
    tags=["metadata"],
    response_model_exclude={
        "result": {
            k.value: {"label_description": True, "result_description": True}
            for k in IndicatorEnum
        },
    },
)
async def metadata_indicators(
    project: ProjectEnum = DEFAULT_PROJECT,
) -> IndicatorMetadataResponse:
    """Get metadata of all indicators."""
    if project == ProjectEnum.all:
        project = None
    return IndicatorMetadataResponse(result=get_indicator_definitions(project=project))


@app.get(
    "/metadata/indicators/{key}",
    tags=["metadata"],
    response_model_exclude={
        "result": {
            k.value: {"label_description": True, "result_description": True}
            for k in IndicatorEnum
        }
    },
)
async def metadata_indicators_by_key(key: IndicatorEnum) -> IndicatorMetadataResponse:
    """Get metadata of an indicator by key."""
    return IndicatorMetadataResponse(
        result={key.value: get_metadata("indicators", hyphen_to_camel(key.value))}
    )


@app.get(
    "/metadata/reports",
    tags=["metadata"],
    response_model_exclude={
        "result": {k.value: {"label_description": True} for k in ReportEnum}
    },
)
async def metadata_reports(
    project: ProjectEnum = DEFAULT_PROJECT,
) -> ReportMetadataResponse:
    """Get metadata of all indicators."""
    if project == ProjectEnum.all:
        project = None
    return ReportMetadataResponse(result=get_report_definitions(project=project))


@app.get(
    "/metadata/reports/{key}",
    tags=["metadata"],
    response_model_exclude={
        "result": {k.value: {"label_description": True} for k in ReportEnum}
    },
)
async def metadata_reports_by_key(key: ReportEnum) -> ReportMetadataResponse:
    """Get metadata of an indicator by key."""
    return ReportMetadataResponse(
        result={key.value: get_metadata("reports", hyphen_to_camel(key.value))}
    )
