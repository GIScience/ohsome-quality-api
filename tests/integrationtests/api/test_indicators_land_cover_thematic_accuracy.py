import json

import pytest
from approvaltests import Options, verify_as_json

from tests.approvaltests_namers import PytestNamer
from tests.conftest import FIXTURE_DIR
from tests.integrationtests.api.test_indicators import (
    RESPONSE_SCHEMA_GEOJSON,
    RESPONSE_SCHEMA_JSON,
)
from tests.integrationtests.utils import oqapi_vcr

ENDPOINT = "/indicators/land-cover-thematic-accuracy"


@pytest.fixture
def mock_db_fetch(monkeypatch):
    async def fetch(*_):
        with open(
            FIXTURE_DIR / "land-cover-thematic-accuracy-db-fetch-results.json", "r"
        ) as file:
            return json.load(file)

    monkeypatch.setattr(
        "ohsome_quality_api.indicators.land_cover_thematic_accuracy.indicator.client.fetch",
        fetch,
    )


@oqapi_vcr.use_cassette
@pytest.mark.parametrize(
    "headers,schema",
    [
        ({"accept": "application/json"}, RESPONSE_SCHEMA_JSON),
        ({"accept": "application/geo+json"}, RESPONSE_SCHEMA_GEOJSON),
    ],
)
def test_multi_class(client, bpolys, headers, schema, mock_db_fetch):
    # corine class parameter is optional (default all corine classes)
    parameters = {"bpolys": bpolys, "topic": "lulc"}
    response = client.post(ENDPOINT, json=parameters, headers=headers)
    assert response.status_code == 200
    assert schema.is_valid(response.json())


@oqapi_vcr.use_cassette
@pytest.mark.parametrize(
    "headers,schema",
    [
        ({"accept": "application/json"}, RESPONSE_SCHEMA_JSON),
        ({"accept": "application/geo+json"}, RESPONSE_SCHEMA_GEOJSON),
    ],
)
def test_single_class(client, bpolys, headers, schema, mock_db_fetch):
    # Corine class 23 are Pastures
    parameters = {"bpolys": bpolys, "topic": "lulc", "corineClass": "23"}
    response = client.post(ENDPOINT, json=parameters, headers=headers)
    assert response.status_code == 200
    assert schema.is_valid(response.json())


def test_invalid_topic(client, bpolys):
    parameters = {"bpolys": bpolys, "topic": "building-count"}
    response = client.post(ENDPOINT, json=parameters)
    assert response.status_code == 422
    verify_as_json(response.json(), options=Options().with_namer(PytestNamer()))


def test_invalid_class(client, bpolys):
    parameters = {"bpolys": bpolys, "topic": "lulc", "corineClass": "1"}
    response = client.post(ENDPOINT, json=parameters)
    assert response.status_code == 422
    verify_as_json(response.json(), options=Options().with_namer(PytestNamer()))
