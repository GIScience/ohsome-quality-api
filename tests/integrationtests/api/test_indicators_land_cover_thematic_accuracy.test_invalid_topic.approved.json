{
    "apiVersion": "1.9.0",
    "detail": [
        {
            "msg": "Invalid combination of indicator and topic: land-cover-thematic-accuracy and key='building-count' name='Buildings (count)' description=\"All buildings as defined by all objects tagged with 'building=*'.\" endpoint='elements' aggregation_type='count' filter='building=* and building!=no and geometry:polygon' indicators=['mapping-saturation', 'currentness', 'attribute-completeness'] projects=[<ProjectEnum.core: 'core'>, <ProjectEnum.bkg: 'bkg'>] source=None ratio_filter=None"
        }
    ],
    "type": "IndicatorTopicCombinationError"
}
